fix(core): detect DI parameters in JIT mode for downleveled ES2015 classes

devversion opened the pull request:
See individual commits.

Fixes #38453

devversion committed:
fix(core): detect DI parameters in JIT mode for downleveled ES2015 classes

In the Angular Package Format, we always shipped UMD bundles and previously even ES5 module output.
With V10, we removed the ES5 module output but kept the UMD ES5 output.

For this, we were able to remove our second TypeScript transpilation. Instead we started only
building ES2015 output and then downleveled it to ES5 UMD for the NPM packages. This worked
as expected but unveiled an issue in the `@angular/core` reflection capabilities.

In JIT mode, Angular determines constructor parameters (for DI) using the `ReflectionCapabilities`. The
reflection capabilities basically read runtime metadata of classes to determine the DI parameters. Such
metadata can be either stored in static class properties like `ctorParameters` or within TypeScript's `design:params`.

If Angular comes across a class that does not have any parameter metadata, it tries to detect if the
given class is actually delegating to an inherited class. It does this naively in JIT by checking if the
stringified class (function in ES5) matches a certain pattern. e.g.

```js
function MatTable() {
  var _this = _super.apply(this, arguments) || this;
```

These patterns are reluctant to changes of the class output. If a class is not recognized properly, the
DI parameters will be assumed empty and the class is **incorrectly** constructed without arguments.

This actually happened as part of v10 now. Since we downlevel ES2015 to ES5 (instead of previously
compiling sources directly to ES5), the class output changed slightly so that Angular no longer detects
it. e.g.

```js
var _this = _super.apply(this, __spread(arguments)) || this;
```

This happens because the ES2015 output will receive an auto-generated constructor if the class
defines class properties. This constructor is then already containing an explicit `super` call.

```js
export class MatTable extends CdkTable {
    constructor() {
        super(...arguments);
        this.disabled = true;
    }
}
```

If we then downlevel this file to ES5 with `--downlevelIteration`, TypeScript adjusts the `super` call so that
the spread operator is no longer used (not supported in ES5). The resulting super call is different to the
super call that would have been emitted if we would directly transpile to ES5. Ultimately, Angular no
longer detects such classes as having an delegate constructor -> and DI breaks.

We fix this by expanding the rather naive RegExp patterns used for the reflection capabilities
so that downleveled pass-through/delegate constructors are properly detected. There is a risk
of a false-positive as we cannot detect whether `__spread` is actually the TypeScript spread
helper, but given the reflection patterns already make lots of assumptions (e.g. that `super` is
actually the superclass, we should be fine making this assumption too. The false-positive would
not result in a broken app, but rather in unnecessary providers being injected (as a noop).

Fixes #38453

JoostK commented on packages/compiler-cli/ngcc/test/host/esm5_host_spec.ts:
```diff
@@ -1497,6 +1497,145 @@ runInEachFileSystem(() => {
+      describe('ES2015 -> ES5: synthesized constructors through TSC downleveling', () => {
+        function getConstructorParameters(
+            constructor: string, mode: 'inlined'|'inlined_with_suffix'|'imported') {
+          let fileHeader = '';
+
+          switch (mode) {
+            case 'imported':
+              fileHeader = `import {__spread} from 'tslib';`;
+              break;
+            case 'inlined':
+              fileHeader =
+                  `var __spread = (this && this.__spread) || function (...args) { /* ... */ }`;
+              break;
+            case 'inlined_with_suffix':
+              fileHeader =
+                  `var __spread$1 = (this && this.__spread$2) || function (...args) { /* ... */ }`;
+              break;
+          }
```
The `fileHeader` isn't actually used here in the below file.

JoostK commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,148 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+    // Check if the `arguments` identifier is wrapped in a call expression that could
+    // potentially correspond to the TypeScript spread helper. We check for the helper
+    // last as that is the most expensive operation here.
+    if (!ts.isCallExpression(argumentsArgument) || !ts.isIdentifier(argumentsArgument.expression) ||
```
The `!ts.isIdentifier(argumentsArgument.expression)` bit is a problem here, as it doesn't allow for namespaced imports (`tslib.__spread`) as it would be in both UMD and CJS. I notice we don't have any of the synthesized ctor tests in the UMD/CJS host tests, and I think it's a good idea to port them over so that we can be confident it works correctly.

(I experimented with this locally to be certain it's currently broken, which is how I noticed that `fileHeader` is not used at all)

devversion replied:
Good catch. Ported the tests over to the UMD host. As a side note: I hate how hard those are to maintain 😄 I wish we would build these "test cases" automatically like I did with the `genrule` here.

JoostK replied:
They are quite a pain to maintain, indeed, but they do allow for testing various forms that could be tedious to replicate.

devversion replied:
Yeah, but I imagine a base of tests is similar (such as the ones I'm porting over to UMD and CJS). There can always still be the advanced custom scenarios. Definitely a larger task; but wanted to mention it as I feel a bit unhappy about this duplication!

JoostK replied:
I feel your pain! In hindsight, I think it would have made sense to use a base test suite of TS code and expectations and run that through the TS compiler and the appropriate hosts. In the integration tests there's several `comileInto*` helpers to do a similar thing, where all we need to maintain is readable TS code.

One downside could be for cases where TS emit changes slightly, in which case we wouldn't have any coverage of the previous emit output. Although we have a similar, though inverse, problem right now, where changes in emit structure are not detected.

JoostK commented on packages/core/src/reflection/reflection_capabilities.ts:
````diff
@@ -17,14 +17,45 @@ import {GetterFn, MethodFn, SetterFn} from './types';
+ * it intends to capture the pattern where existing constructors have been downleveled from
+ * ES2015 to ES5 using TypeScript w/ downlevel iteration. e.g.
+ *
+ *  * ```
````
```suggestion
 * ```
```

JoostK commented on packages/core/test/BUILD.bazel:
```diff
@@ -15,13 +15,31 @@ circular_dependency_test(
     deps = ["//packages/core/testing"],
 )
 
+genrule(
```
This is really clever!

devversion commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,170 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+
+    const left = expression.left;
+    if (isBinaryExpr(left, ts.SyntaxKind.AmpersandAmpersandToken)) {
+      return isSuperNotNull(left.left) && this.isSuperApplyCall(left.right);
```
One worthy note for reviewer: Previously, this function _always_ asserted that there is the `super !== null` check before `super` is invoked. This assertion won't be generated in the downleveled ES2015 output as the constructor seems explicitly called to TS when it processes the ES2015 output to ES5.

I've updated the logic to skip the `super !== null` check if there is no binary expression, but if there is, then we still enforce that this is actually the expected `super !== null` check. I'm not sure of any risk here as we even ensure that there are no parameters on the containing constructor (for which we perform this check).

alxhub replied:
One thing that would be very helpful both for reviewers and future readers of this code would be to include examples of the shape of code being matched by these functions/conditionals.

devversion replied:
I had such examples for all of these functions; but there are quite a lot of these and it felt even worse repeating the JSDoc examples multiple times (resulting in inconsistency and difficulity to maintain). I'll check if I can add a smaller one here.

devversion replied:
@alxhub added back more concrete examples to all functions (with concrete descriptions). It's more to maintain; but should be okay given that the synthetic delegate output isn't changing too often.

alxhub replied:
Awesome, thanks!

JoostK commented on packages/compiler-cli/ngcc/test/host/umd_host_spec.ts:
```diff
@@ -2910,6 +2910,259 @@ runInEachFileSystem(() => {
     });
 
+
+    describe('TS -> ES5: synthesized constructors', () => {
```
Consider moving these inside the `getConstructorParameters` `describe` as is the case in ES5.

Also, WDYT about also porting this to the CJS tests?

devversion replied:
Both 👍 

Also updated tests to only define one of the test helper functions per `describe` (reducing at least a bit of duplication!)

JoostK reviewed (approved):
LGTM for ngcc changes, with some comments.

devversion commented on packages/compiler-cli/ngcc/test/host/esm5_host_spec.ts:
```diff
@@ -1497,6 +1497,146 @@ runInEachFileSystem(() => {
+            constructor: string, mode: 'inlined'|'inlined_with_suffix'|'imported') {
+          let fileHeader = '';
+
+          switch (mode) {
```
yeah, I like that. adding

devversion committed:
fix(ngcc): detect synthesized delegate constructors for downleveled ES2015 classes

Similarly to the change we landed in the `@angular/core` reflection
capabilities, we need to make sure that ngcc can detect pass-through
delegate constructors for classes using downleveled ES2015 output.

More details can be found in the preceding commit, and in the issue
outlining the problem: #38453.

Fixes #38453.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,170 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+    if (isBinaryExpr(left, ts.SyntaxKind.AmpersandAmpersandToken)) {
+      return isSuperNotNull(left.left) && this.isSuperApplyCall(left.right);
+    }
+    return this.isSuperApplyCall(left);
```
Since this is an either-or case, put it in an `else` block with the case above.

devversion replied:
I was recently told that it would be more clearer to not nest code if we return in the first block already. I don't feel strongly about that; but generally would also prefer reduced nesting. 

Will make the change though.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,170 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+    if (!ts.isCallExpression(expression) || expression.arguments.length !== 2) return false;
+
+    const targetFn = expression.expression;
+    if (!ts.isPropertyAccessExpression(targetFn)) return false;
```
Feel free to combine multiple guards into a single `if` block.

devversion replied:
This is code that just moved around, and I didn't want to make a lot of unrelated changes/cleanups.

devversion commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,170 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+   * @returns true if the constructor appears to have been synthesized
+   */
+  private isSynthesizedConstructor(constructor: ts.FunctionDeclaration): boolean {
+    if (!constructor.body) return false;
```
Sure! I use these conventions too, but this is just moved code that already existed. Seems like something to cleanup separately.

alxhub replied:
Ah, that's okay with me.

devversion commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -352,6 +352,170 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
     return ts.isBlock(classDeclarationParent) ? Array.from(classDeclarationParent.statements) : [];
   }
+
+  ///////////// Host Private Helpers /////////////
```
Yeah, they could, but I moved the code explicitly into the class so that I can rely on the host for detecting the spread helper, without having to pass through the host X times.

alxhub replied:
👍

devversion committed:
fixup! fix(ngcc): detect synthesized delegate constructors for downleveled ES2015 classes

Add better descriptions to helper methods; stylistic preference feedback

devversion commented:
@alxhub Addressed all of the feedback that I think it applicable for this PR / in scope. Please have another look. 

Note that I did not clean up existing code that just moved around because that seems out of scope for this PR. A separate cleanup PR could be sent to cleanup the existing code in the ngcc host. Happy to help with that.
